js/foam/dao/AuthenticatedDAO.js

```diff
@@ -0,0 +1,210 @@
+  // for a new method to be added, but not authenticated.
+  properties: [
+    {
+      name: 'model',
+      required: true
+    },
+    {
+      name: 'delegate',
+      required: true
+    },
+    {
+      name: 'authenticator',
+      documentation: 'An instance of $$DOC{ref:"foam.dao.Authenticator"} ' +
+          'that describes how to authenticate each DAO method.',
+      required: true
+    },
+  ],
+
+  methods: [
+    function adelegateFind_(id) {
+      return function(ret) {
+        this.delegate.find(id, {
+          put: ret,
+          error: function() { ret(null); }
+        });
+      }.bind(this);
+    },
+    function put(obj, sink, opt_X) {
+      // On put, ask the authenticator to decorate our object.
+      // It might return null, indicating that the object is owned by another
+      // user, and this put should be denied.
+      var self = this;
+      var principal = opt_X && opt_X.principal;
+      if (!principal) {
+        sink && sink.error && sink.error('Illegal put.');
+        return;
+      }
+
+      aseq(
+        // Always return null if ID is not set; this is a new object.
+        obj.id ? this.adelegateFind_(obj.id) : aconstant(null),
+        function(ret, old) {
+          self.authenticator.massageForPut(ret, principal, old, obj);
+        },
+        function(ret, massaged) {
+          if (massaged === null) {
+            sink && sink.error && sink.error('Illegal put.');
+          } else {
+            self.delegate.put(massaged, {
+              error: function(e) {
+                sink && sink.error && sink.error(e);
+                ret();
+              },
+              put: function(postPut) {
+                self.authenticator.massageForRead(function(massaged) {
+                  if (massaged) {
+                    sink && sink.put && sink.put(massaged);
+                  } else {
+                    sink && sink.error && sink.error('Illegal put.');
+                  }
+                  ret();
+                }, principal, postPut);
+              }
+            });
+          }
+        }
+      )(function(massaged) {
+      });
+    },
+    function remove(id_or_obj, sink, opt_X) {
+      // On remove, we retrieve the original object from the delegate, then ask
+      // the authenticator whether the removal is allowed. If the user sent us a
+      // whole object instead of an ID, we don't trust any part of it except the
+      // ID.
+      var self = this;
+      var principal = opt_X && opt_X.principal;
+      if (!principal) {
+        sink && sink.error && sink.error('Illegal remove.');
+        return;
+      }
+
+      var id = id_or_obj.id || id_or_obj;
+
+      aseq(
+        this.adelegateFind_(id),
+        function(ret, old) {
+          self.authenticator.shouldAllowRemove(ret, principal, old);
+        }
+      )(function(allowed) {
+        if (allowed) {
+          self.delegate.remove(id, sink);
```

Comment:
Removes will read the object back out with sink.remove(obj);  You should decorate the sink here and do massageForRead on the remove callback.